runtime/src/accounts_db.rs

```diff
@@ -2795,13 +3224,29 @@ impl AccountsDB {
             hashes,
             storage_finder,
             write_version_producer,
+            use_cache,
         );
         store_accounts_time.stop();
         self.stats
             .store_accounts
             .fetch_add(store_accounts_time.as_us(), Ordering::Relaxed);
         let mut update_index_time = Measure::start("update_index");
-        let reclaims = self.update_index(slot, infos, accounts);
+
+        // If the cache was flushed, then becasue `update_index` occurs
+        // after the accounts are stored, all reads after this point
+        // will know to not check the cache anymore
+        let mut reclaims = self.update_index(slot, infos, accounts);
+
+        // For each updated account, `reclaims` should only have at most one
+        // item (if the account was previously updated in this slot).
+        // filter out the cached reclaims as those don't actually map
+        // to anything that needs to be cleaned in the backing storage
+        // entries
+        reclaims.retain(|(_, r)| r.store_id != Self::cache_store_id());
```

Comment:
nits: move reclaims.retain(...) to the else of `if use_cache {`? (no point to do `.retain` if it's `.is_empty()`.)

Reply:
ah yup, good call

Reply:
so the assertion 

```
if use_cache {
   assert!(reclaims.is_empty());
}
```

is actually only true because it comes *after* the `reclaims.retain(|(_, r)| r.store_id != Self::cache_store_id());`, i.e. those should filter out the *one* entry in the reclaims (recall if you store into the same slot, the length of the reclaims is at most 1).